source/extensions/filters/network/dubbo_proxy/metadata.h

```diff
@@ -0,0 +1,128 @@
+
+#pragma once
+
+#include <map>
+#include <string>
+
+#include "common/http/header_map_impl.h"
+
+#include "extensions/filters/network/dubbo_proxy/protocol.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace DubboProxy {
+
+class MessageMetadata {
+public:
+  struct ParameterValue {
+    ParameterValue(uint32_t index, const std::string& value) : index_(index), value_(value) {}
+
+    uint32_t index_;
+    std::string value_;
+  };
+  typedef std::unordered_map<uint32_t, ParameterValue> ParameterValueMap;
+  typedef std::unique_ptr<ParameterValueMap> ParameterValueMapPtr;
+
+  typedef std::unique_ptr<Http::HeaderMapImpl> HeaderMapPtr;
+
+  void setServiceName(const std::string& name) { service_name_ = name; }
+  const std::string& service_name() const { return service_name_; }
+
+  void setMethodName(const std::string& name) { method_name_ = name; }
+  const absl::optional<std::string>& method_name() const { return method_name_; }
+
+  void setServiceVersion(const std::string& version) { service_version_ = version; }
+  const absl::optional<std::string>& service_version() const { return service_version_; }
+
+  void setServiceGroup(const std::string& group) { group_ = group; }
+  const absl::optional<std::string>& service_group() const { return group_; }
+
+  void setMessageType(MessageType type) { message_type_ = type; }
+  MessageType message_type() const { return message_type_; }
+
+  void setRequestId(int64_t id) { request_id_ = id; }
+  int64_t request_id() const { return request_id_; }
+
+  void setSerializationType(SerializationType type) { serialization_type_ = type; }
+  SerializationType serialization_type() const { return serialization_type_; }
+
+  void setTwoWayFlag(bool two_way) { is_two_way_ = two_way; }
+  bool is_two_way() const { return is_two_way_; }
+
+  void setEventFlag(bool is_event) { is_event_ = is_event; }
+  bool is_event() const { return is_event_; }
+
+  void setResponseStatus(ResponseStatus status) { response_status_ = status; }
+  const absl::optional<ResponseStatus>& response_status() const { return response_status_; }
+
+  void addParameterValue(uint32_t index, const std::string& value) {
+    assignParameterIfNeed();
+    parameter_map_->emplace(index, ParameterValue(index, value));
+  }
+  const ParameterValue* getParameterValue(uint32_t index) const {
+    if (parameter_map_) {
+      auto itor = parameter_map_->find(index);
+      if (itor != parameter_map_->end()) {
+        return &itor->second;
+      }
+    }
+
+    return nullptr;
+  }
+  bool hasParameters() const { return parameter_map_ != nullptr; }
+  const ParameterValueMap& parameters() {
+    assignParameterIfNeed();
+    return *parameter_map_;
+  }
+
+  bool hasHeaders() const { return headers_ != nullptr; }
+  const Http::HeaderMap& headers() {
+    assignHeaderIfNeed();
```

Comment:
probably just ASSERT header is not nullptr and make this method const? You shouldn't access this when hasHeaders return false. (same for parameters)

Reply:
fixed